### DISCUSSION TITLE
Closing connection with magic sql

### DISCUSSION THREAD

**Discussion by dominijk** [duckdb/duckdb#7092] [Q&A]
After some trial and error I've found that to close a connection using ipython-sql you need to go through below steps, is this normal behaviour. Should i need to restart the kernel

1. Close connection using `%sql --close duckdb::////url `
3. Restart the kernel

Is there a duckdb command that would close and release the lock without shutting down the kernel?

**Comment by paultiq** [comment]
> After some trial and error I've found that to close a connection using ipython-sql you need to go through below steps, is this normal behaviour. Should i need to restart the kernel
> 
> 
> 
> 1. Close connection using `%sql --close duckdb::////url `
> 
> 3. Restart the kernel
> 
> 
> 
> Is there a duckdb command that would close and release the lock without shutting down the kernel? 

I think what you want to do is create a new connection. I was trying to figure out something similar, and in a recent [thread in discord](https://discord.com/channels/909674491309850675/921100786098901042/1091331941858562078) ended up with the following:

The key part here is passing a new 'engine' to %sql. This is just meant to illustrate how you can manage the connection yourself, outside jupysql.

```
`%load_ext sql 
import duckdb
import duckdb_engine
from sqlalchemy.engine import create_engine
con = duckdb.connect()
conn_ddb_lambda = lambda x: duckdb_engine.ConnectionWrapper(con)
engine = create_engine("duckdb:///:memory:", creator=conn_ddb_lambda)
%sql engine

%sql pragma version
```

_FYI, as a result of that thread, I ended up writing a separate %dql magic [magic_duckdb](https://pypi.org/project/magic-duckdb/)_

**Reply by dominijk** [reply]
Magic duck looks great, thanks for your work on that. In my example i want to use the same connection (having closed it) between two different ipython / jupyter notebooks in the same session. So, close db connection in notebook 1 open notebook 2 and start new connection to same db

**Reply by paultiq** [reply]
>  close db connection in notebook 1 open notebook 2 and start new connection to same db

For either library, you could create the connection and close it outside the %magic. I often do this because I can use the connection directly. 
- in %sql, you could use the original code I provided to do this. 
- in %dql, use `%dql -co conn`

To do it all within the %magic: 
```
# Install
%pip install magic_duckdb --upgrade
%load_ext magic_duckdb
```

Notebook 1
```
%dql -cn myfile.db
%dql CREATE TABLE IF NOT EXISTS xyz as SELECT * FROM range(10)
%dql --close
```

Notebook 2
```
%dql -cn myfile.db SELECT * from xyz
# connection persists until --close or kernel terminates.
```

Note, this wouldn't work with :memory: connections, since the data won't be persisted across processes.

**Comment by Mause** [comment]
This is probably an issue in duckdb-engine rather than duckdb itself, I'll try to find time to have a look later this week 😅
